contrib/src/main/scala/akka/stream/contrib/SourceGen.scala

```diff
@@ -0,0 +1,141 @@
+
+    val generateUnfoldFlowGraphStageLogic = (shape: FanOutShape2[O, S, E]) => new UnfoldFlowGraphStageLogic[O, S, E](shape, seed) {
+      setHandler(nextElem, new InHandler {
+        override def onPush() = {
+          val o = grab(nextElem)
+          unfoldWith(o) match {
+            case None => completeStage()
+            case Some((s, e)) => {
+              pending = s
+              push(output, e)
+              pushedToCycle = false
+            }
+          }
+        }
+      })
+    }
+
+    unfoldFlowGraph(new FanOut2unfoldingStage(generateUnfoldFlowGraphStageLogic), flow)
+  }
+
+  private[akka] def unfoldFlowGraph[E, S, O, M](
+    fanOut2Stage: GraphStage[FanOutShape2[O, S, E]],
+    flow:         Graph[FlowShape[S, O], M]
+  ): Source[E, M] = Source.fromGraph(GraphDSL.create(flow) {
+    implicit b =>
+      {
+        f =>
+          {
+            import GraphDSL.Implicits._
+
+            val fo2 = b.add(fanOut2Stage)
+            fo2.out0 ~> f ~> fo2.in
+            SourceShape(fo2.out1)
+          }
+      }
+  })
+
+  private[akka] abstract class UnfoldFlowGraphStageLogic[O, S, E] private[stream] (shape: FanOutShape2[O, S, E], seed: S) extends GraphStageLogic(shape) {
+
+    lazy val timeout = Duration.fromNanos(ConfigFactory.load().getDuration("akka.stream.contrib.unfold-flow-timeout").toNanos)
+    val feedback = shape.out0
+    val output = shape.out1
+    val nextElem = shape.in
+
+    var pending: S = seed
+    var pushedToCycle = false
+
+    setHandler(feedback, new OutHandler {
+      override def onPull() = if (!pushedToCycle && isAvailable(output)) {
+        push(feedback, pending)
+        pending = null.asInstanceOf[S]
+        pushedToCycle = true
+      }
+
+      override def onDownstreamFinish() = {
+        // Do Nothing until `timeout` to try and intercept completion as downstream,
+        // but cancel stream after timeout if inlet is not closed to prevent deadlock.
+        materializer.scheduleOnce(timeout, new Runnable {
+          override def run() = {
+            getAsyncCallback[Unit] { _ =>
+              if (!isClosed(nextElem)) {
+                failStage(new IllegalStateException(s"unfoldFlow source's inner flow canceled only upstream, while downstream remain available for $timeout"))
+              }
+            }.invoke(())
+          }
+        })
+      }
+    })
+
+    setHandler(output, new OutHandler {
+      override def onPull() = {
+        pull(nextElem)
+        if (!pushedToCycle && isAvailable(feedback)) {
+          push(feedback, pending)
+          pending = null.asInstanceOf[S]
+          pushedToCycle = true
+        }
+      }
+    })
+  }
+
+  private[akka] class FanOut2unfoldingStage[O, S, E] private[stream] (generateGraphStageLogic: FanOutShape2[O, S, E] => UnfoldFlowGraphStageLogic[O, S, E]) extends GraphStage[FanOutShape2[O, S, E]] {
```

Comment:
@drewhk should I not restrict with `private[akka]`/`private[stream]`?
https://github.com/akka/akka/issues/20543